go/store/prolly/tree/patch_generator.go

```diff
@@ -0,0 +1,533 @@
+// Copyright 2025 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tree
+
+import (
+	"context"
+	"fmt"
+	"io"
+)
+
+// A Patch represents a single change being applied to a tree.
+// If Level == 0, then this is a change to a single key, and |KeyBelowStart| and |SubtreeCount| are ignored.
+// If Level > 0, then To is either an address or nil, and this patch represents a change to the range (KeyBelowStart, EndKey].
+// An address indicates that the every row in the provided range should be replaced by the rows found by loading the address.
+// A nil address indicates that every row in the provided range has been removed.
+type Patch struct {
+	From          Item
+	KeyBelowStart Item
+	EndKey        Item
+	To            Item
+	SubtreeCount  uint64
+	Level         int
+}
+
+func newAddedPatch(previousKey, key Item, to Item, subtreeCount uint64, level int) Patch {
+	return Patch{
+		From:          nil,
+		KeyBelowStart: previousKey,
+		EndKey:        key,
+		To:            to,
+		SubtreeCount:  subtreeCount,
+		Level:         level,
+	}
+}
+
+func newModifiedPatch(previousKey, key Item, from, to Item, subtreeCount uint64, level int) Patch {
+	return Patch{
+		From:          from,
+		KeyBelowStart: previousKey,
+		EndKey:        key,
+		To:            to,
+		SubtreeCount:  subtreeCount,
+		Level:         level,
+	}
+}
+
+func newRemovedPatch(previousKey, key Item, from Item, level int) Patch {
+	return Patch{
+		From:          from,
+		KeyBelowStart: previousKey,
+		EndKey:        key,
+		To:            nil,
+		SubtreeCount:  0,
+		Level:         level,
+	}
+}
+
+func newLeafPatch(key Item, from, to Item) Patch {
+	return Patch{
+		From:   from,
+		EndKey: key,
+		To:     to,
+		Level:  0,
+	}
+}
+
+type PatchIter interface {
+	NextPatch(ctx context.Context) Patch
```

Comment:
Nit: I find it weird to take `context` and not return an `error`. `context` almost always means I/O or cancelable or something. I realize there are other iterators in `tree` and friends that already do this. Thoughts?